### DISCUSSION TITLE
Problem in the adding a new element.

### DISCUSSION THREAD

**Discussion by Diego-planckian1** [iqm-finland/KQCircuits#76] [Q&A] [answered]
Hi,

I'm novel in the use of KQCircuit in the development mode.

I tried to add the tutorial "simple cross" as new element, but, even if I can see the code in the KLayout macros, the simple cross element does not appear in the element library in the KLayout section.

I press KQCircuit>Reload Libraries, but it seems to not change anything.

I'm working on ubuntu 22 and I installed the latest version of Klayout. Could be a problem of software version or anything else?

Thanks

**Comment by qpavsmi** [comment]
Hi, thanks for trying out KQCircuits!

Did you place your `.py` code for the "simple cross" under the `klayout_package/python/kqcircuits/elements` directory?

**Reply by Diego-planckian1** [reply]
Hi! yes, I follow all the instructions

**Reply by qpavsmi** [reply]
Following works for me -> Create new file `klayout_package/python/kqcircuits/elements/my_box.py` with following content:

```
from kqcircuits.elements.element import Element
from kqcircuits.pya_resolver import pya


class MyBox(Element):
    """Just a box that shows up in Elements Library
    """
    def build(self):
        box = pya.DBox(pya.DPoint(-1,-1),pya.DPoint(1,1))
        self.cell.shapes(self.get_layer("base_metal_gap_wo_grid")).insert(box)
```
Then loading KLayout I can see "My Box" entry in the Element Library

![image](https://github.com/iqm-finland/KQCircuits/assets/109152229/3b824c69-26f3-4835-8bfd-c7ec30cd28f8)

Important thing is that the code is within the class that is a subclass of `Element`.

If you replicate these instructions do you see "My Box" listed in "Element Library" on your device?

**Comment by Diego-planckian1** [comment]
I still not see that:
![Screenshot from 2023-12-13 22-11-41](https://github.com/iqm-finland/KQCircuits/assets/150356602/c5417f26-5d94-48ff-84d1-6f3e30d1ad77)
![Screenshot from 2023-12-13 22-12-16](https://github.com/iqm-finland/KQCircuits/assets/150356602/a11f34a3-eb12-4ea1-b5ac-83d097cf08c9)


![Screenshot from 2023-12-13 22-14-38](https://github.com/iqm-finland/KQCircuits/assets/150356602/333bfd63-656c-41d5-a0f6-e25ef8880e46)

**Reply by qpavsmi** [reply]
Hmm this is unexpected behaviour. Here are some troubleshooting that comes to my mind at this stage:

1. Try looking through every library, not only Element Library. See if for some reason the listing is not in alphabetic order.
2. Check out the `File > Log Viewer` for errors or warnings. Normally it looks empty but maybe it will complain for you about something related to this problem
![image](https://github.com/iqm-finland/KQCircuits/assets/109152229/4c7891aa-ebab-4124-b9ce-bcf987b8e83f)
3. Try starting a KLayout GUI session that opens your element immediately on startup. You can do this with this command (cd to KQCircuits source):

  `klayout -e -rx -rm util/create_element_from_path.py -rd element_path=kqcircuits\elements\simple_cross.py`

Should display the element as expected. If you get the 3rd part working then at least this gives a way for you to debug your elements

**Reply by Diego-planckian1** [reply]
1. The problem stands even for other libraries and I don't find the  simple cross despite the alphabetic order.

2. Log viewer seems to be okay if I simply reload the libraries: ![Screenshot from 2023-12-14 11-18-55](https://github.com/iqm-finland/KQCircuits/assets/150356602/f0f5bf78-afe6-4010-bd16-f28a1a68881f)

3.If i run the command you told me I get this:

![Screenshot from 2023-12-14 11-15-17](https://github.com/iqm-finland/KQCircuits/assets/150356602/42156bc7-fbd5-45ae-8fa0-431c2a6dbdfb)
![Screenshot from 2023-12-14 11-15-30](https://github.com/iqm-finland/KQCircuits/assets/150356602/76c65b1e-eff0-4587-929a-b20de71caf2d)
![Screenshot from 2023-12-14 11-20-55](https://github.com/iqm-finland/KQCircuits/assets/150356602/ff8ab09d-1f23-4c1a-b911-5aa2f05c8f6e)


Thanks a lot for your help !

**Reply by Diego-planckian1** [reply]
Could be a problem  of versions?
![Screenshot from 2023-12-14 11-54-12](https://github.com/iqm-finland/KQCircuits/assets/150356602/10667ddd-5756-4d72-a468-219b69bb2b5a)

**Reply by qpavsmi** [reply]
The weird thing is that it doesn't seem that KLayout even recognizes the new python file as a module, yet in the Macro development utility the file is listed like any other. Here are some other things to try:

# 1
Open Macro development (F5) in KLayout, create a new macro, and paste the following:
```
from kqcircuits.elements.my_box import MyBox # replace with SimpleCross
from kqcircuits.klayout_view import KLayoutView

view = KLayoutView(current=True)
view.insert_cell(MyBox, pya.DTrans(pya.DVector(0, 0)), {})
```
Then "run script form current tab" (shift + F5) or play button with exclamation mark. The expected result is that the element will be placed in the layout that is currently open in the gui (might have to press `F2` to zoom to full view and `*` to show entire cell hierarchy)

# 2
Check file permissions of your new files, so that they're consistent with other file's permissions

# 3
Open python in the terminal and try `from kqcircuits.elements.simple_cross import SimpleCross`, will you get errors? If you try to import our elements such as `from kqcircuits.elements.smooth_capacitor import SmoothCapacitor` do you get errors from that?

# 4
You might have better luck with using our user package system, instructions here: https://iqm-finland.github.io/KQCircuits/salt.html#setting-up-a-user-package-directory This allows to load in elements and chips from the directory of your choice into KQCircuits, instead of placing everything in the KQC source directory.

Python version 3.10 should be fine, you can try updating but I doubt that's the problem

Maybe you could also verify the following things:
- You have not installed KQCircuits in KLayout's internal "Salt package manager", or if you have in the past, you have removed that package
- After cloning KQCircuits, you have run `python setup_within_klayout.py` successfully
- You have `.klayout` folder in your home directory and in there there is a `python` folder and inside symbolic links called `kqcircuits` and `kqcircuits_scripts`, both leading to the cloned KQCircuits code directory
- You have installed KQCircuits as standalone. When you write `pip list` in the terminal you can see listing for `kqcircuits` and it shows that it also lists the directory to the cloned KQCircuits code

If all of this fails I can't really come up with anything else except to try to reinstall everything related to KLayout if you can

**Reply by Diego-planckian1** [reply]
I didn't understand the 4th point... salt package must be installed also if I use standalone mode?

**Reply by qpavsmi** [reply]
Salt package shouldn't be installed. But the "Add user package" and "Remove user package" features are available if you have KQCircuits installed following the Developer guide. You can try creating some folder anywhere you want, then "Add user package", give it a name and make it point to your directory. Maybe that way the code you put there will show up in the Library

**Comment by iqmtestd** [comment]
Hi,

I'm also using Ubuntu 22. You should not have this problem. It should be simple. :) It looks like a configuration issue to me.

Always use venv!

But to be sure, I've tried with a fresh config and a fresh KQCircuits clone directory:
```
python -m venv env-test
source env-test/bin/activate

cd KQCircuits/
pip install -e klayout_package/python/[docs,tests]
python ./setup_within_klayout.py

klayout -e -rx -rm util/create_element_from_path.py -rd element_path=kqcircuits/elements/finger_capacitor_square.py

vim klayout_package/python/kqcircuits/elements/my_box.py
klayout -e -rx -rm util/create_element_from_path.py -rd element_path=kqcircuits/elements/my_box.py
```

This works for me.

**Reply by Diego-planckian1** [reply]
Thank you for your help. All the terminal lines you wrote also work for me, excepting the last one! Again KLayout does not find my_box.py...

**Reply by iqmtestd** [reply]
If you see `finger_capacitor_square` but not the new `my_box` element then you've accidentally set up two (or more) environments. This is [possble](https://iqm-finland.github.io/KQCircuits/developer/setup.html#secondary-install) but you should remember to prefix `klayout` with `KLAYOUT_HOME=~/.klayout_alt/my_other_KQCircuits_dir ` (or such like).

**Reply by iqmtestd** [reply]
If one environment is enough then use this quick fix: `rm -rf ~/.klayout ~/.klayout_alt` and run `setup_within_klayout` in your chosen KQCircuits directory again.

**Reply by Diego-planckian1** [reply]
I've done everything that you said... still "ModuleNotFoundError: No module named 'kqcircuits.elements.my_box'"

**Reply by iqmtestd** [reply]
You do have a working environment. After all you could run `finger_capacitor_square.py` in KLayout. Either you have an other directory configured or you do not have the `my_box.py` there, or maybe some typo somewhere.

You could try to modify `finger_capacitor_square.py` and see it changing (or just put `import pdb; pdb.set_trace()` as the first line of `build()`)  to prove to yourself that you are using the right environment. Then any other correctly written file in the same directory will work the same way.

Alternatively, you could run `ls -l ~/.klayout/python` to see where your configured KQCircuits folder really is.

**Reply by iqmtestd** [reply]
A perhaps even simpler alternative solution is to create a new linux user.

Log in with the new user,  clone the sources (`git clone https://github.com/iqm-finland/KQCircuits.git`) and run the commands in my [first post](https://github.com/iqm-finland/KQCircuits/discussions/76#discussioncomment-7853551).

**Reply by Diego-planckian1** [reply, marked as answer]
Hi, after a reset of my user on ubuntu it works! It was a problem of environment as you said! Thank you very much for your help!!!

**Reply by qpavsmi** [reply]
Happy to hear! Hope you'll find KQCircuits useful!
